lingvo/tasks/asr/ctc_model.py

```diff
@@ -34,212 +38,234 @@ class CTCModel(base_model.BaseTask):
-      outputs['encoder_outputs'] = rnn_out
-      outputs['encoder_outputs_padding'] = rnn_padding
-      return outputs
-
-  def Inference(self):
-    subgraphs = {}
-    with tf.name_scope('inference'):
-      subgraphs['default'] = self._InferenceSubgraph_Default()
-    return subgraphs
-
-  def _InferenceSubgraph_Default(self):
+  def CreateDecoderMetrics(self):
+    base_metrics = {
+      "wer": metrics.AverageMetric(),
+      "cer": metrics.AverageMetric(),
+      "num_samples_in_batch": metrics.AverageMetric()
+    }
+    return base_metrics
+
+  @classmethod
+  def FPropMeta(cls, params, *args, **kwargs):
+    raise NotImplementedError("No FPropMeta available.")
+
+  def _FrontendAndEncoderFProp(self, theta, input_batch_src):
     p = self.params
-    with tf.name_scope('default'):
-      wav_bytes = tf.placeholder(dtype=tf.string, name='wav')
-      frontend = self.frontend if p.frontend else None
-      if not frontend:
-        # No custom frontend. Instantiate the default.
-        frontend_p = asr_frontend.MelAsrFrontend.Params()
-        frontend = frontend_p.Instantiate()
-
-      # Decode the wave bytes and use the explicit frontend.
-      unused_sample_rate, audio = audio_lib.DecodeWav(wav_bytes)
-      # Doing this the "kaldi" way, not the pytorch audio way
-      audio *= 32768
-      # Remove channel dimension, since we have a single channel.
-      audio = tf.squeeze(audio, axis=1)
-      # Add batch.
-      audio = tf.expand_dims(audio, axis=0)
+    in_shape = tf.shape(input_batch_src.src_inputs)
+
+    if p.frontend:
+      input_batch_src = self.frontend.FProp(theta.frontend, input_batch_src)
+
+    if p.input_stacking_tpl:
+      inputs = tf.squeeze(input_batch_src.src_inputs, [-1])
+      rnn_padding = tf.expand_dims(input_batch_src.paddings, -1)
+      inputs, rnn_padding = self.input_stacking.FProp(inputs, rnn_padding)
       input_batch_src = py_utils.NestedMap(
-          src_inputs=audio, paddings=tf.zeros_like(audio))
-      input_batch_src = frontend.FPropDefaultTheta(input_batch_src)
+        src_inputs=tf.expand_dims(inputs, -1),
+        paddings=tf.squeeze(rnn_padding, [-1])
+      )
 
-      encoder_outputs = self.FPropDefaultTheta() # _FProp()
-      decoder_outputs = self.decoder.BeamSearchDecode(encoder_outputs)
-      topk = self._GetTopK(decoder_outputs)
+    encoder_out = self.encoder.FProp(theta.encoder, input_batch_src)
 
-      feeds = {'wav': wav_bytes}
-      fetches = {
-          'hypotheses': topk.decoded,
-          'scores': topk.scores,
-          'src_frames': input_batch_src.src_inputs,
-          'encoder_frames': encoder_outputs.encoded
-      }
+    encoded, padding = self.project_to_vocab_size(encoder_out.encoded, encoder_out.padding)
+    outputs = py_utils.NestedMap(encoded=encoded, padding=padding)
+    return outputs
 
-      return fetches, feeds
+  def _DecodeCTC(self, output_batch):
+    # swap row 0 and row 73 because decoder assumes blank is at 0,
```

Comment:
BTW, my new tokenizer always assumes blank is at row 0 because the WFST decoder depends upon that being the case. It's rather silly to have a magic constant like this, but it seems the machine learning world has settled upon 0 being the index for blank.

Reply:
Stale comment. No longer hard code the idx of blank token, we read the index of blank token from params. So, it should work with whatever index your tokenizer has.